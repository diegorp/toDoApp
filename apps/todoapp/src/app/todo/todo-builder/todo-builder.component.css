:root {
    --muted-color: mix(white, #424242, 70%);
}

.wrapper {
    position: relative;
}

input {
  background: none;
  color: var(--muted-color);
  font-size: 18px;
  padding: 10px 10px 10px 5px;
  display: block;
  width: 100%;
  border: none;
  border-radius: 0;
  border-bottom: 1px solid var(--muted-color);
}

input:focus {
    outline: none;
}

input:focus ~ label,
input.active ~ label {
    top: -14px;
    font-size: 12px;
    color: #2196F3;
}

input:focus ~ .bar:before {
    width: 100%;
}

label {
    color: #9f9f9f;
    font-size: 16px;
    font-weight: normal;
    position: absolute;
    pointer-events: none;
    left: 5px;
    top: 10px;
    transition: 300ms ease all;
}

.bar {
    position: relative;
    display: block;
    width: 100%;
}

.bar:before {
    content: '';
    height: 2px;
    width: 0;
    bottom: 0px;
    position: absolute;
    background: #2196F3;
    transition: 300ms ease all;
    left: 0%;
}

button {
    background: none;
    border: none;
}

button:before {
    content: '+';
    height: 32px;
    width: 32px;
    top: 0;
    right: 0;
    position: absolute;
    background-color: #2196F3;
    color: #f2f2f2;
    padding: 4px 8px;
    font-size: 2em;
    border-radius: 50%;
}