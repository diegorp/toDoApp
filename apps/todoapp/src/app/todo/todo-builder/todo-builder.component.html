<div class="wrapper">
    <input type="text" [class.active]="toDo" [value]="toDo" (input)="updateToDo($event)">
    <span class="bar"></span>
    <label>Add new TODO</label>
    <button (click)="addToDo()"></button>
</div>