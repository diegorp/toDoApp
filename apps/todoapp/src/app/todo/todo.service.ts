import { Injectable } from '@angular/core';
import { Observable, Subject } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ToDoService {
  
  private readonly toDo = new Subject<string>();

  getToDo(): Observable<string> {
    return this.toDo.asObservable();
  }

  addToDo(newToDo: string) {
    this.toDo.next(newToDo);
  }
}
