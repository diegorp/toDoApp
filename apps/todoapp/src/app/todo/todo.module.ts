import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TodoItemComponent } from './todo-item/todo-item.component';
import { TodoListComponent } from './todo-list/todo-list.component';
import { TodoBuilderComponent } from './todo-builder/todo-builder.component';



@NgModule({
  declarations: [
    TodoItemComponent,
    TodoListComponent,
    TodoBuilderComponent
  ],
  imports: [
    CommonModule
  ],
  exports: [
    TodoListComponent,
    TodoBuilderComponent
  ]
})
export class TodoModule { }
