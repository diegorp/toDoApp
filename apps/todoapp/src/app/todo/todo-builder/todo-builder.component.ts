import { Component } from '@angular/core';
import { ToDoService } from '../todo.service';

@Component({
  selector: 'todoapp-todo-builder',
  templateUrl: './todo-builder.component.html',
  styleUrls: ['./todo-builder.component.css']
})
export class TodoBuilderComponent {

  toDo = '';

  constructor(private readonly toDoService: ToDoService) {}

  updateToDo(possibleToDo: any) {
    this.toDo = possibleToDo?.target?.value;
  }

  addToDo() {
    this.toDoService.addToDo(this.toDo);
    this.toDo = '';
  }
}
