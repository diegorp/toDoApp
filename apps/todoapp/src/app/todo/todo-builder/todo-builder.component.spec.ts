import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { ToDoService } from '../todo.service';

import { TodoBuilderComponent } from './todo-builder.component';

jest.mock('../todo.service');

describe('TodoBuilderComponent', () => {
  let component: TodoBuilderComponent;
  let fixture: ComponentFixture<TodoBuilderComponent>;
  let toDoService: ToDoService;
  
  const toDoText = 'Get some coffee';

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ TodoBuilderComponent ],
      providers: [ ToDoService ],
      schemas: [ NO_ERRORS_SCHEMA ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(TodoBuilderComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    toDoService = TestBed.inject(ToDoService);
    jest.spyOn(toDoService, 'addToDo');
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should update the to do', () => {
    const inputElement = fixture.debugElement.query(By.css('input'));
    inputElement.nativeElement.value = toDoText;
    inputElement.nativeElement.dispatchEvent(new Event('input'));
    expect(component.toDo).toEqual(toDoText);
  });

  it('should send the value in the input field to the toDoService after clicking on add button', () => {
    const inputElement = fixture.debugElement.query(By.css('input'));
    inputElement.nativeElement.value = toDoText;
    inputElement.nativeElement.dispatchEvent(new Event('input'));
    const buttonElement = fixture.debugElement.query(By.css('button'));
    buttonElement.nativeElement.click();
    expect(toDoService.addToDo).toHaveBeenCalledWith(toDoText);
  });

  it('should render the component', () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });
});
