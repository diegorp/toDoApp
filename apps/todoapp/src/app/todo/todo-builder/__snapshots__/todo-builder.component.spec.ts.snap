// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoBuilderComponent should render the component 1`] = `
<div
  id="root3"
>
  <div
    class="wrapper"
  >
    <input
      type="text"
    />
    <span
      class="bar"
    />
    <label>
      Add new TODO
    </label>
    <button />
  </div>
</div>
`;
